:sparkles: ec2: Add support for userdata privacy (#1490)
kubernetes-sigs/cluster-api-provider-aws: docs/userdata-privacy.md

# Userdata Privacy
Cluster API Provider AWS bootstraps EC2 instances to create and join Kubernetes clusters using instance user data.
Because Kubernetes clusters are secured using TLS using multiple Certificate Authorities, these are generated by
Cluster API and injected into the user data. It is important to note that without the configuring of host firewalls, processes can
retrieve instance userdata from http://169.254.169.254/latest/api/token

## How Cluster API secures TLS secrets

> In v1alpha3, by default, Cluster API Provider AWS will use [AWS Secrets Manager](https://aws.amazon.com/secrets-manager/)

Reviewer: v0.5.x / v1alpha3

> to store the userdata, with a boot script added to the userdata to download the userdata, and additionally immediately delete it.

Reviewer: ```suggestion
to store the actual userdata in encrypted form. The normal unencrypted userdata uses a boot script to download the encrypted userdata secret using a time-limited URL, immediately deletes it from AWS Secrets Manager, and then executes it.
```

ish

Cluster API Provider AWS also stores the secret ARN in the AWSMachine spec, and will delete the secret if it isn't already deleted and
the machine has registered successfully against the workload cluster API server as a node.
Cluster API Provider AWS will also attempt deletion of the secret if the AWSMachine is otherwise deleted or the EC2 instance

> is terminated.

Reviewer: ... or failed

# Userdata Privacy

> Cluster API Provider AWS bootstraps EC2 instances to create and join Kubernetes clusters using instance user data.

Reviewer 2: Leave space above

Because Kubernetes clusters are secured using TLS using multiple Certificate Authorities, these are generated by
Cluster API and injected into the user data. It is important to note that without the configuring of host firewalls, processes can
retrieve instance userdata from http://169.254.169.254/latest/api/token

## How Cluster API secures TLS secrets

In 0.5.x/v1alpha3, by default, Cluster API Provider AWS will use [AWS Secrets Manager](https://aws.amazon.com/secrets-manager/)
to store the actual userdata in encrypted form. The normal unencrypted userdata uses a boot script to download the encrypted userdata secret

> using a instance profile permissions, immediately deletes it from AWS Secrets Manager, and then executes it.

Reviewer: ```suggestion
using instance profile permissions, immediately deletes it from AWS Secrets Manager, and then executes it.
```

> To avoid guessing keys in the AWS Secrets Manager key-value store and to prevent collisions, the key is an of encoding the

Reviewer: ```suggestion
To avoid guessing keys in the AWS Secrets Manager key-value store and to prevent collisions, the key is an encoding the
```
